[RHCLOUD-18651] EAN to OrgId migration

MikelAlejoBR opened the pull request:
Modifies the tenancy table by adding the "org_id" column, and attempting a non-hard-failing translation from current EANs to OrgIds.

**NOTE:** https://github.com/RedHatInsights/sources-api-go/pull/175 needs to be merged first, and then, this PR needs to be rebased.

## Links

[[RHCLOUD-18651]](https://issues.redhat.com/browse/RHCLOUD-18651)

MikelAlejoBR committed:
feature: migration to support "OrgId"

lindgrenj6 commented on db/migrations/20220330120000_org_id_support.go:
```diff
@@ -0,0 +1,175 @@
+		ID: "20220330120000",
+		Migrate: func(db *gorm.DB) error {
+			logging.Log.Info(`Migration "20220330120000" started`)
+			defer logging.Log.Info(`Migration "20220330120000" ended`)
```
Can we give this a "more pretty" name? Something like `Migration of external tenants to org_id's started|ended`

MikelAlejoBR replied:
Done

lindgrenj6 commented on db/migrations/20220330120000_org_id_support.go:
```diff
@@ -0,0 +1,175 @@
+					Updates(map[string]interface{}{
+						"external_tenant": nil,
+						"org_id":          result.OrgID,
+					})
```
So we **definitely** do not want to set `external_tenant` to nil during this migration. We are going to need to keep it around until none of our clients need it, we'll just drop the entire column at a future date.

MikelAlejoBR replied:
I thought we would convert EBS and OrgIds back and forth on the fly, in the case we received an EBS account number for a tenant that does have a corresponding OrgId. However it is true that your suggestion adds that extra safety in the case that the translation service is down.

Done!

lindgrenj6 commented on db/migrations/20220330120000_org_id_support.go:
```diff
@@ -0,0 +1,175 @@
+
+			for _, result := range results {
+				if result.Err != nil {
+					logging.Log.Errorf(`[external_tenant: %s] could not translate to "org_id": %s`, *result.EAN, err)
```
Maybe need a `continue` after the log.

MikelAlejoBR replied:
Great catch :tropical_fish: , done!

lindgrenj6 commented on db/migrations/20220330120000_org_id_support.go:
```diff
@@ -0,0 +1,175 @@
+
+			return nil
+		},
+		Rollback: func(db *gorm.DB) error {
```
I think the rollback for this function can just be using the migrator and dropping the column since we're going to leave the `external_tenant` field as is.

MikelAlejoBR replied:
Done!

lindgrenj6 reviewed (commented):
Good first pass - I think this is basically all we'll need for the migration piece.

MikelAlejoBR committed:
refactor: improve log messages

MikelAlejoBR committed:
refactor: skip loop if the EAN could not be converted

MikelAlejoBR committed:
refactor: don't delete the EAN from the database

MikelAlejoBR committed:
refactor: on rollback, simply drop the orgId column

MikelAlejoBR committed:
refactor: don't overwrite the EAN number

MikelAlejoBR commented on deploy/clowdapp.yaml:
```diff
@@ -200,3 +202,7 @@ parameters:
+- description: Host for the EAN to OrgId translator
+  name: TENANT_TRANSLATOR_HOST
+  required: true
+  value: '?????????'
```
Do you know where I can find the tenant translator host for this one? I tried finding it on the documents to give the variable a sane default but... :sweat:

lpichler commented on db/migrations/20220330120000_org_id_support.go:
```diff
@@ -0,0 +1,131 @@
+			translator := tenantid.NewTranslator(config.Get().TenantTranslatorHost)
+			results, err := translator.EANsToOrgIDs(context.Background(), ebsAccountNumbers)
+			if err != nil {
+				return err
```
What will follow when connection to translator service will fail ? 
In which state will be pod - do we need to do anything in this case ?

MikelAlejoBR replied:
This migration is supposed to be a "best effort" migration, since we're proactively adopting OrgIds. This means that if the service fails for whichever reason, we don't have to worry too much, since we are expecting some EANs to be untranslatable.

lpichler commented on db/migrations/20220330120000_org_id_support.go:
```diff
@@ -0,0 +1,131 @@
+// translation for an EBS account number is acceptable.
+//
+// It also adds a comment to the "external_tenant" column, clarifying that it refers to "EBS account numbers".
+func OrgIdSupport() *gormigrate.Migration {
```
(1)I would prefer whether this name would describe more what it is doing(+name file).
"AddOrgIDtoTenants"
"Translate EBS account number to ORG ID" and maybe it can also correspond sort of with "start" and "end" log messages 

(2) Also maybe it would be good to don't mix structural and data migrations.

Let me know what do you think about it.

MikelAlejoBR replied:
> (1)I would prefer whether this name would describe more what it is doing(+name file). "AddOrgIDtoTenants" "Translate EBS account number to ORG ID"

Sure. I am very bad at naming functions so that's why I end up writing comments on them to give them more context.

> (2) Also maybe it would be good to don't mix structural and data migrations.

Why is this? In this case the migration doesn't fail if the data migration does not succeed, because it's a "since we're at it, let's try populating the column too" kind of thing. That's why I conceived it as part of the same migration.

lpichler replied:
(2) It is just my opinion - I think there is too much things in one migration:
- add column + add commenT
- translation 

it could be two migrations - reason is just to be more clear what is happening in one migration. I see one migration as one unit where less things (reasonable) as possible - and then then we don't need go to start back when anythings fail.
Also naming of migrations is more unambiguous without hiding additions in migration - you can see on first look what is going inside migration.
+ when we would test data migration it would be also more clear what is being tested.

yes and now I understand more role of the translation operation.

I don't insist  on it-  I leave it  up to you - I wanted to see your opinion - If you like we can adopt it as "rule" migration.

I would like also hear @lindgrenj6's opinion

MikelAlejoBR replied:
> it could be two migrations - reason is just to be more clear what is happening in one migration. I see one migration as one unit where less things (reasonable) as possible - and then then we don't need go to start back when anythings fail.
> Also naming of migrations is more unambiguous without hiding additions in migration - you can see on first look what is going inside migration.

Okay that's a good point. After refactoring it is true that it is easier to follow. I submitted the commit.

lpichler replied:
great, thanks!

MikelAlejoBR committed:
refactor: rename migration for clarity

MikelAlejoBR committed:
refactor: split the orgId migration in two for clarity

lpichler commented on db/migrations/20220404150000_translate_ebs_org_ids.go:
```diff
@@ -0,0 +1,130 @@
+// translation for an EBS account number is acceptable.
+//
+// It also adds a comment to the "external_tenant" column, clarifying that it refers to "EBS account numbers".
+func TranslateEbsAccountNumbersToOrgIds() *gormigrate.Migration {
```
this method  is not part of `migrations.migrationsCollection`

lpichler commented on db/migrations/20220404150000_translate_ebs_org_ids.go:
```diff
@@ -0,0 +1,130 @@
+			err := db.Transaction(func(tx *gorm.DB) error {
+				// Add the comment to the "external_tenant" column manually, since doing it in the struct with a "gorm"
+				// tag makes "gormigrate" try to create the already existing column.
+				err := tx.
```
I think this is in previous migration already :)

lindgrenj6 reviewed (approved):
(this is leftover from yesterday, I just didn't click submit)

LGTM, going to merge this next week after we do the cutover.

MikelAlejoBR committed:
refactor: add scheme to the tenant translator env var

Clowder/Bonfire replace the "host" variables with just hosts, without
the scheme. This makes us have to also specify the scheme for the URL.

MikelAlejoBR committed:
refactor: skip the tenant translation if there is nothing to translate
